### PR TITLE
Fix the issue of executing SQL Query without using Prepared Statement

### PR DESCRIPTION
**Pull request by UdeshAthukorala**
### Proposed changes in this pull request

Fix the issue of executing the SQL Query without using Prepared Statement.

Related Issue : https://github.com/wso2/product-is/issues/11748

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java [hunk 1]
```diff
@@ -957,13 +957,13 @@
             "SELECT SCOPES.SCOPE_ID, SCOPES.NAME, SCOPES.DISPLAY_NAME, SCOPES.DESCRIPTION, " +
                     "SCOPEBINDINGS.SCOPE_BINDING, SCOPEBINDINGS.BINDING_TYPE FROM IDN_OAUTH2_SCOPE SCOPES " +
                     "LEFT JOIN IDN_OAUTH2_SCOPE_BINDING SCOPEBINDINGS ON SCOPES.SCOPE_ID=SCOPEBINDINGS.SCOPE_ID " +
-                    "WHERE SCOPES.TENANT_ID=? AND SCOPES.SCOPE_TYPE=? AND SCOPES.NAME IN (?)";
+                    "WHERE SCOPES.TENANT_ID=? AND SCOPES.SCOPE_TYPE=? AND SCOPES.NAME IN (_SCOPE_LIST_)";
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java:960]: let's use a constant for `_SCOPE_LIST_` and reuse it in other places

Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1603/commits/9a22c8f2df57eb1ad51c7db8e3b7e63b971a03e1

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/Oauth2ScopeConstants.java [hunk 2]
```diff
@@ -110,6 +110,7 @@ public String toString() {
         public static final String LIMIT = "limit";
         public static final String OFFSET = "offset";
         public static final String SCOPE_TYPE = "scope_type";
+        public static final String SCOPE_LIST = "_SCOPE_LIST_";
```
Comment (UdeshAthukorala) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/Oauth2ScopeConstants.java:113]:
```suggestion
        public static final String SCOPE_LIST_PLACEHOLDER = "_SCOPE_LIST_";
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthScopeDAOImpl.java [hunk 3]
```diff
@@ -261,19 +267,25 @@ public void addScope(Scope scope, int tenantID) throws IdentityOAuth2ScopeExcept
         }
 
         List<String> requestedScopeList = Arrays.asList(requestedScopes.split("\\s+"));
-        String sqlIN = requestedScopeList.stream().map(x -> String.valueOf(x))
-                .collect(Collectors.joining("\', \'", "(\'", "\')"));
+        final StringBuilder sb = new StringBuilder(String.join(", ",
+                Collections.nCopies(requestedScopeList.size(), "?")));
 
-        sql = sql.replace("(?)", sqlIN);
+        sql = sql.replace("(" + Oauth2ScopeConstants.SQLPlaceholders.SCOPE_LIST + ")", "(" + sb + ")");
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthScopeDAOImpl.java:273]:
```suggestion
        sql = sql.replace(Oauth2ScopeConstants.SQLPlaceholders.SCOPE_LIST, sb);
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthScopeDAOImpl.java [hunk 4]
```diff
@@ -261,19 +267,25 @@ public void addScope(Scope scope, int tenantID) throws IdentityOAuth2ScopeExcept
...
         Set<Scope> scopes = new HashSet<>();
         Map<Integer, Scope> scopeMap = new HashMap<>();
 
         try (Connection conn = IdentityDatabaseUtil.getDBConnection(false)) {
             try (PreparedStatement ps = conn.prepareStatement(sql)) {
                 ps.setInt(1, tenantID);
+                int scopeIndex = 2;
                 if (!includeOIDCScopes) {
                     ps.setString(2, Oauth2ScopeConstants.SCOPE_TYPE_OAUTH2);
+                    scopeIndex = 3;
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthScopeDAOImpl.java:284]:
```suggestion
                    scopeIndex++;
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java [hunk 5]
```diff
@@ -957,13 +957,15 @@
             "SELECT SCOPES.SCOPE_ID, SCOPES.NAME, SCOPES.DISPLAY_NAME, SCOPES.DESCRIPTION, " +
                     "SCOPEBINDINGS.SCOPE_BINDING, SCOPEBINDINGS.BINDING_TYPE FROM IDN_OAUTH2_SCOPE SCOPES " +
                     "LEFT JOIN IDN_OAUTH2_SCOPE_BINDING SCOPEBINDINGS ON SCOPES.SCOPE_ID=SCOPEBINDINGS.SCOPE_ID " +
-                    "WHERE SCOPES.TENANT_ID=? AND SCOPES.SCOPE_TYPE=? AND SCOPES.NAME IN (?)";
+                    "WHERE SCOPES.TENANT_ID=? AND SCOPES.SCOPE_TYPE=? AND SCOPES.NAME IN (" +
+                    Oauth2ScopeConstants.SQLPlaceholders.SCOPE_LIST + ")";
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java:961]: use static imports

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java [hunk 6]
```diff
@@ -957,13 +957,15 @@
...
 
     public static final String RETRIEVE_REQUESTED_ALL_SCOPES_WITHOUT_SCOPE_TYPE =
             "SELECT SCOPES.SCOPE_ID, SCOPES.NAME, SCOPES.DISPLAY_NAME, SCOPES.DESCRIPTION, " +
                     "SCOPEBINDINGS.SCOPE_BINDING, SCOPEBINDINGS.BINDING_TYPE FROM IDN_OAUTH2_SCOPE SCOPES " +
                     "LEFT JOIN IDN_OAUTH2_SCOPE_BINDING SCOPEBINDINGS ON SCOPES.SCOPE_ID=SCOPEBINDINGS.SCOPE_ID " +
-                    "WHERE SCOPES.TENANT_ID=? AND SCOPES.NAME IN (?)";
+                    "WHERE SCOPES.TENANT_ID=? AND SCOPES.NAME IN (" +
+                    Oauth2ScopeConstants.SQLPlaceholders.SCOPE_LIST + ")";
```
Comment (omindu) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java:968]: use static imports

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthScopeDAOImpl.java (hunk 3 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthScopeDAOImpl.java]: Addressd with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1603/commits/94aae9a3b3fb20f88adc5bba70f73dcbdd149cf8

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthScopeDAOImpl.java (hunk 4 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/OAuthScopeDAOImpl.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1603/commits/94aae9a3b3fb20f88adc5bba70f73dcbdd149cf8

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java (hunk 5 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1603/commits/94aae9a3b3fb20f88adc5bba70f73dcbdd149cf8

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java (hunk 6 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/SQLQueries.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1603/commits/94aae9a3b3fb20f88adc5bba70f73dcbdd149cf8

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/Oauth2ScopeConstants.java (hunk 2 above)
Comment (UdeshAthukorala) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/Oauth2ScopeConstants.java]: Addressed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1603/commits/94aae9a3b3fb20f88adc5bba70f73dcbdd149cf8

**Review by omindu — APPROVED** [review]

**Pull request merged**